### PR TITLE
Remove/assert casts

### PR DESCRIPTION
**Pull request by elazarg**
Remove a cast to Any and some others, and assert-guard many casts.

The patch for each file is independent of the others. I think the only non-immediate change is in semanal.py.

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -1276,6 +1276,7 @@ def check_lvalue(self, lvalue: Lvalue) -> Tuple[Type, IndexExpr, Var]:
                 assert isinstance(inferred, Var)
             else:
                 m = cast(MemberExpr, lvalue)
```
Comment (ddfisher) [inline mypy/checker.py]: I think you forgot to remove this cast?

File: mypy/checkexpr.py [hunk 2]
```diff
@@ -609,8 +612,9 @@ def apply_inferred_arguments(self, callee_type: CallableType,
         # Apply the inferred types to the function type. In this case the
         # return type must be CallableType, since we give the right number of type
         # arguments.
-        return cast(CallableType, self.apply_generic_arguments(callee_type,
-                                                           inferred_args, context))
+        res = self.apply_generic_arguments(callee_type, inferred_args, context)
+        assert isinstance(res, CallableType)
+        return cast(CallableType, res)
```
Comment (ddfisher) [inline mypy/checkexpr.py]: Is this cast necessary?

File: mypy/checkmember.py [hunk 3]
```diff
@@ -249,6 +249,7 @@ def analyze_var(name: str, var: Var, itype: Instance, info: TypeInfo, node: Cont
                 if var.is_property:
                     # A property cannot have an overloaded type => the cast
                     # is fine.
+                    assert isinstance(signature, CallableType)
                     return cast(CallableType, signature).ret_type
```
Comment (ddfisher) [inline mypy/checkmember.py]: I checked this cast -- we can remove it.

File: mypy/semanal.py [hunk 4]
```diff
@@ -323,34 +315,36 @@ def prepare_method_signature(self, func: FuncDef) -> None:
...
     def update_function_type_variables(self, defn: FuncDef) -> None:
         """Make any type variables in the signature of defn explicit.
 
         Update the signature of defn to contain type variable definitions
         if defn is generic.
         """
-        if defn.type:
-            functype = cast(CallableType, defn.type)
-            typevars = self.infer_type_variables(functype)
+        if isinstance(defn.type, CallableType):
```
Comment (ddfisher) [inline mypy/semanal.py]: I think this might be cleaner with an assert instead.  This changes the meaning a bit.

File: mypy/semanal.py [hunk 5]
```diff
@@ -385,6 +381,7 @@ def find_type_variables_in_type(
             name = type.name
             node = self.lookup_qualified(name, type)
             if node and node.kind == UNBOUND_TVAR:
+                assert isinstance(node.node, TypeVarExpr)
                 result.append((name, cast(TypeVarExpr, node.node)))
```
Comment (ddfisher) [inline mypy/semanal.py]: This cast can be removed.

File: mypy/semanal.py [hunk 6]
```diff
@@ -2351,6 +2352,7 @@ def lookup(self, name: str, ctx: Context) -> SymbolTableNode:
         # 5. Builtins
         b = self.globals.get('__builtins__', None)
         if b:
+            assert isinstance(b.node, MypyFile)
             table = cast(MypyFile, b.node).names
```
Comment (ddfisher) [inline mypy/semanal.py]: This cast can likely be removed.

**Comment by elazarg** [comment]
Thanks for the review. I left some casts thinking they may serve as a warning sign to the reader, but of course `assert`s can also be grepped.

File: mypy/semanal.py [hunk 7]
```diff
@@ -2406,23 +2407,20 @@ def lookup_qualified(self, name: str, ctx: Context) -> SymbolTableNode:
 
     def builtin_type(self, fully_qualified_name: str) -> Instance:
         node = self.lookup_fully_qualified(fully_qualified_name)
-        info = cast(TypeInfo, node.node)
-        return Instance(info, [])
+        assert isinstance(node.node, TypeInfo)
+        return Instance(node.node, [])
 
     def lookup_fully_qualified(self, name: str) -> SymbolTableNode:
```
Comment (elazarg) [inline mypy/semanal.py]: This refactoring is not really related to the casts, but why not?

**Comment by elazarg** [comment]
@ddfisher I think it's done. Please notice the change in how errors are reported in `apply_generic_arguments`, the removal of `apply_generic_arguments2`, and the refactoring of `lookup_fully_qualified`.

File: mypy/semanal.py (hunk 7 above)
Comment (ddfisher) [inline mypy/semanal.py]: Actually, this PR is pretty big already (and is not without risk of breakage -- changing casts to asserts potentially changes runtime behavior), so I'd prefer to keep it as focused as possible.  If you'd like to make this refactor, could you submit it as a separate small PR?

**Comment by elazarg** [comment]
@ddfisher I don't understand your comment about `apply_generic_arguments`. Sorry. Can you elaborate?

File: mypy/subtypes.py [hunk 8]
```diff
@@ -333,7 +333,7 @@ def unify_generic_callable(type: CallableType, target: CallableType,
         return None
     msg = messages.temp_message_builder()
     applied = mypy.applytype.apply_generic_arguments(type, inferred_vars, msg, context=target)
-    if msg.is_errors() or not isinstance(applied, CallableType):
+    if applied is None or msg.is_errors():
```
Comment (elazarg) [inline mypy/subtypes.py]: `not isinstance(applied, CallableType)` before if and only if `applied is None` now.
`not msg.iserrors()` also implies `not isinstance(applied, AnyType) ` and therefore `isinstance(applied, CallableType)`.
Am I wrong?

**Comment by gvanrossum** [comment]
Could you split this up in smaller portions, e.g. one PR per file?

**Comment by elazarg** [comment]
Working on it

**Comment by elazarg** [comment]
Changes to 9 files remain in this PR, but each file has 1-2 uncomplicated changes.

**Comment by gvanrossum** [comment]
Waiting for @ddfisher to finish the review.

File: mypy/checker.py [hunk 9]
```diff
@@ -1314,6 +1314,7 @@ def check_lvalue(self, lvalue: Lvalue) -> Tuple[Type, IndexExpr, Var]:
...
+                assert isinstance(m, MemberExpr)
```
Comment (rwbarton) [inline mypy/checker.py]: Shouldn't these lines be reversed to get rid of a `cast`?

File: mypy/typeanal.py [hunk 10]
```diff
@@ -291,6 +291,7 @@ def anal_var_defs(self, var_defs: List[TypeVarDef]) -> List[TypeVarDef]:
 
     def builtin_type(self, fully_qualified_name: str, args: List[Type] = None) -> Instance:
         node = self.lookup_fqn_func(fully_qualified_name)
+        assert isinstance(node.node, TypeInfo)
         info = cast(TypeInfo, node.node)
```
Comment (rwbarton) [inline mypy/typeanal.py]: You should be able to get rid of this cast (and could get rid of `info` entirely).

**Comment by ddfisher** [comment]
This looks good now, thanks!  It looks like there's a small merge conflict to fix, though.

**Comment by elazarg** [comment]
Fixed. Thanks.

**Comment by gvanrossum** [comment]
Thanks! I liked all of these improvements. Death to casts!
